File: hooks/tts-notify.py
#!/usr/bin/env python3
"""Cross-platform hook wrapper: tts-notify.

On Unix: delegates to tts-notify.sh.
On Windows: native Python implementation.

Claude Code Hook Protocol (PostToolUse):
  Receives JSON on stdin: {"tool_name": "...", "tool_input": {...}, ...}
  Exit 0: always (notification hook, never blocks)

FAILURE POLICY: FAIL-OPEN
"""
import json
import os
import shlex
import sys
import tempfile
import time
import urllib.request
from pathlib import Path

HOOK_DIR = Path(__file__).parent


def main() -> None:
    if sys.platform != "win32":
        script = HOOK_DIR / "tts-notify.sh"
        os.execv("/bin/bash", ["bash", str(script)] + sys.argv[1:])

    # Windows: native Python implementation
    threshold = int(os.environ.get("SPELLBOOK_TTS_THRESHOLD", "30"))
    mcp_port = os.environ.get("SPELLBOOK_MCP_PORT", "8765")
    mcp_host = os.environ.get("SPELLBOOK_MCP_HOST", "127.0.0.1")
    speak_url = f"http://{mcp_host}:{mcp_port}/api/speak"

    blacklist = {
        "AskUserQuestion", "TodoRead", "TodoWrite",
        "TaskCreate", "TaskUpdate", "TaskGet", "TaskList",
    }

    stdin_data = sys.stdin.read()
    if not stdin_data:
        sys.exit(0)

    try:
        data = json.loads(stdin_data)
    except json.JSONDecodeError:
        sys.exit(0)

    tool_name = data.get("tool_name", "")
    if tool_name in blacklist:
        sys.exit(0)

    tool_use_id = data.get("tool_use_id", "")
    if not tool_use_id:
        sys.exit(0)

    start_file = Path(tempfile.gettempdir()) / f"claude-tool-start-{tool_use_id}"
    try:
        start_ts = int(start_file.read_text().strip())
        start_file.unlink()
    except (FileNotFoundError, ValueError):
        sys.exit(0)

    elapsed = int(time.time()) - start_ts
    if elapsed < threshold:
        sys.exit(0)

    # Build message
    cwd = data.get("cwd", "")
    project = os.path.basename(cwd) if cwd else "unknown"

    # Detail extraction (matches tts-notify.sh logic)
    inp = data.get("tool_input") or {}
    detail = ""
    if tool_name == "Bash":
        cmd = inp.get("command", "")
        if cmd:
            try:
                parts_cmd = shlex.split(cmd)
            except ValueError:
                parts_cmd = cmd.split()
            detail = parts_cmd[0].split("/")[-1] if parts_cmd else ""
    elif tool_name == "Task":
        detail = inp.get("description", "")[:40]

    parts = [project, tool_name]
    if detail:
        parts.append(detail)
    parts.append("finished")
    message = " ".join(parts)

    # Send to MCP server
    try:
        payload = json.dumps({"text": message}).encode()
        req = urllib.request.Request(
            speak_url, data=payload,
            headers={"Content-Type": "application/json"},
            method="POST",
        )
        urllib.request.urlopen(req, timeout=10)
    except Exception:
        pass  # Fail-open

    sys.exit(0)


if __name__ == "__main__":
    main()

File: hooks/nim/src/tts_notify.nim
## tts_notify - PostToolUse hook for TTS notifications.
##
## Phase: PostToolUse (catch-all, no matcher)
## Timeout: 15s, async
## Failure policy: FAIL-OPEN
##
## Sends POST to MCP server's /api/speak endpoint when a tool
## execution exceeds the configured threshold (default 30s).
## This hook benefits most from Nim: the shell version makes
## 3 separate python3 invocations; Nim does it all natively.

import std/[json, os, strutils, options]
import hooklib

const BLACKLIST = ["AskUserQuestion", "TodoRead", "TodoWrite",
                   "TaskCreate", "TaskUpdate", "TaskGet", "TaskList"]

proc main() =
  let input = readStdinJson()
  if input.isNil: quit(0)

  let toolName = input.getOrDefault("tool_name").getStr("")
  let toolUseId = input.getOrDefault("tool_use_id").getStr("")
  let cwd = input.getOrDefault("cwd").getStr("")

  # Check blacklist
  if toolName in BLACKLIST: quit(0)

  # Validate tool_use_id
  if toolUseId.len == 0: quit(0)
  let sanitized = sanitizeToolUseId(toolUseId)
  if sanitized.isNone: quit(0)

  # Read start timestamp
  let startFile = "/tmp/claude-tool-start-" & sanitized.get
  if not fileExists(startFile): quit(0)

  var startTs: int64
  try:
    let content = readFile(startFile).strip()
    startTs = parseBiggestInt(content)
  except ValueError, IOError:
    quit(0)

  # Delete start file
  try:
    removeFile(startFile)
  except OSError:
    discard

  # Check threshold
  let threshold = parseInt(getEnvOr("SPELLBOOK_TTS_THRESHOLD", "30"))
  let now = unixTimestamp()
  let elapsed = now - startTs
  if elapsed < threshold: quit(0)

  # Build announcement message
  let project = if cwd.len > 0: cwd.lastPathPart else: "unknown"

  var detail = ""
  let toolInput = input{"tool_input"}
  if toolName == "Bash" and not toolInput.isNil:
    let cmd = toolInput.getOrDefault("command").getStr("")
    if cmd.len > 0:
      let tokens = shellTokenize(cmd)
      if tokens.len > 0:
        detail = tokens[0].lastPathPart  # Strip path prefix
  elif toolName == "Task" and not toolInput.isNil:
    detail = toolInput.getOrDefault("description").getStr("")
    if detail.len > 40:
      detail = detail[0..<40]

  var parts = @[project, toolName]
  if detail.len > 0:
    parts.add(detail)
  parts.add("finished")
  let message = parts.join(" ")

  if message.len == 0: quit(0)

  # Send to MCP server
  let host = getEnvOr("SPELLBOOK_MCP_HOST", "127.0.0.1")
  let port = parseInt(getEnvOr("SPELLBOOK_MCP_PORT", "8765"))
  speakVia(host, port, message)

  quit(0)

main()
